platform/qt_location_service.cpp

```diff
@@ -0,0 +1,160 @@
+#include "platform/qt_location_service.hpp"
+
+#include "base/logging.hpp"
+#include "base/macros.hpp"
+
+#include "std/target_os.hpp"
+
+#include <QGeoPositionInfoSource>
+
+namespace
+{
+static location::GpsInfo gpsInfoFromQGeoPositionInfo(QGeoPositionInfo const & i, location::TLocationSource source)
+{
+  location::GpsInfo info;
+  info.m_source = source;
+
+  info.m_latitude = i.coordinate().latitude();
+  info.m_longitude = i.coordinate().longitude();
+  info.m_timestamp = static_cast<double>(i.timestamp().toSecsSinceEpoch());
+
+  if (i.hasAttribute(QGeoPositionInfo::HorizontalAccuracy))
+    info.m_horizontalAccuracy = static_cast<double>(i.attribute(QGeoPositionInfo::HorizontalAccuracy));
+
+  if (i.hasAttribute(QGeoPositionInfo::VerticalAccuracy))
+    info.m_verticalAccuracy = static_cast<double>(i.attribute(QGeoPositionInfo::VerticalAccuracy));
+
+  if (i.hasAttribute(QGeoPositionInfo::Direction))
+    info.m_bearing = static_cast<double>(i.attribute(QGeoPositionInfo::Direction));
+
+  if (i.hasAttribute(QGeoPositionInfo::GroundSpeed))
+    info.m_speedMpS = static_cast<double>(i.attribute(QGeoPositionInfo::GroundSpeed));
+
+  return info;
+}
+
+static location::TLocationError tLocationErrorFromQGeoPositionInfoError(QGeoPositionInfoSource::Error error)
+{
+  location::TLocationError result = location::TLocationError::ENotSupported;
+  switch(error)
+  {
+    case QGeoPositionInfoSource::AccessError: result = location::TLocationError::EDenied; break;
+    case QGeoPositionInfoSource::ClosedError: result = location::TLocationError::EGPSIsOff; break;
+    case QGeoPositionInfoSource::NoError: result = location::TLocationError::ENoError; break;
+    case QGeoPositionInfoSource::UpdateTimeoutError: result = location::TLocationError::ETimeout; break;
+    case QGeoPositionInfoSource::UnknownSourceError: result = location::TLocationError::EUnknown; break;
+    default: break;
+  }
+  return result;
+}
+
+static location::TLocationSource qStringToTLocationSource(QString const & sourceName)
+{
+  if ("geoclue2" == sourceName)
+    return location::TLocationSource::EGeoClue2;
+
+  return location::TLocationSource::EUndefined;
+}
+};
+
+
+QtLocationService::QtLocationService(location::LocationObserver & observer, std::string const & sourceName) : LocationService(observer)
+{
+  QVariantMap params;
+  params["desktopId"] = "OrganicMaps";
+  m_positionSource = QGeoPositionInfoSource::createSource(QString::fromStdString(sourceName), params, this);
+
+  if (!m_positionSource)
+  {
+    LOG(LWARNING, ("Failed to acquire QGeoPositionInfoSource from ", sourceName));
+    return;
+  }
+
+  if (!connect(m_positionSource, &QGeoPositionInfoSource::positionUpdated,
+               this, &QtLocationService::OnLocationUpdate))
+  {
+    LOG(LERROR, ("Failed to connect the signal:", "positionUpdated"));
```

Comment:
`CHECK(false, (...));`
Here and below. Crash should be immediate for developers' errors, in debug and in production.

Reply:
It is a developer error if this fails.